### DISCUSSION TITLE
Debugging new OAuth provider

### DISCUSSION THREAD

**Discussion by yannickgloster** [nextauthjs/next-auth#4520] [Help]
### Question 💬

I'm currently trying to use a new implementation of OAuth for Challengermode (https://www.challengermode.com/developers/docs/concepts#personal-access-token) with next-auth and I'm trying to figure out whether the issue lies with next-auth or with Challengermode.

I'm currently able to get successfully to the OAuth page for Challengermode but I am immediately presented with a callback error
```
error: {
    message: 'expected 200 OK, got: 400 Bad Request',
 ...
 name: 'OAuthCallbackError',
  code: undefined
  ```

From what I can tell, this error comes from one of two things, the request to get the token after having consented with OAuth is failing or the fact that there currently isn't a user info URL yet is causing the issue. 

Is there any other way to debug this?

### How to reproduce ☕️

```
{
      id: "challengermode",
      name: "Challengermode",
      type: "oauth",
      issuer: "https://challengermode.com/",
      authorization: {
        url: "https://www.challengermode.com/oauth/authorize",
        params: {
          scope: "offline_access",
        }, // openid
      },
      jwks_endpoint: "https://challengermode.com/.well-known/jwks",
      clientId: ".....",
      checks: ["pkce"],
      client: {
        token_endpoint_auth_method: "none",
      },
      userinfo: null,
      token: {
        url: "https://www.challengermode.com/oauth/token",
        // https://github.com/nextauthjs/next-auth/issues/1846
      },
      profile(profile, tokens) {
        console.log(tokens);
        console.log(profile);
        return {
          id: "test",
        };
      },
      ```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by mrr11k** [comment]
Hi, 
did you solve it somehow?

**Reply by yannickgloster** [reply]
Hey, I was able to work with the Challengermode Devs to write the provider code that worked.

```
{
    id: "challengermode",
    name: "Challengermode",
    type: "oauth",
    authorization: "https://challengermode.com/oauth/authorize",
    checks: ["pkce"],
    token: {
      url: "https://challengermode.com/oauth/token",
      async request(context) {
        var urlParams = new URLSearchParams();
        urlParams.append("grant_type", "authorization_code");
        urlParams.append("code_verifier", context.checks.code_verifier);
        urlParams.append("code", context.params.code);
        urlParams.append("client_id", context.client.client_id);
        urlParams.append("redirect_uri", "http://localhost:3000/api/auth/callback/challengermode%22);

        var response = await fetch(
          "https://challengermode.com/oauth/token",
          {
            method: "POST",
            body: urlParams
          }
        );
        return { "tokens": await response.json() };
      }
    },
    userinfo: "https://publicapi.challengermode.com/mk1/v1/me",
    profile(profile) {
      return {
        id: profile.id,
        name: profile.username,
        image: profile.picture
      };
    },
}
```

**Reply by kart2190** [reply]
Thank you for this. Have been driving myself crazy for two days trying to access params and code_verifier in the same place!!!!

**Comment by yalegria** [comment]
I guess a more general question is how do you debug connections to external auth providers. I am currently trying to debug a connection that first I dont think get established, meaning I dont know if I am hitting the server. Is there a way to debug this?
